fusillade/utils/retry.py

```diff
@@ -0,0 +1,305 @@
+                            tx.logger.warning("Timed out retrying %r.", tx)
+                            raise
+                        else:
+                            tx.logger.debug("Sleeping %f seconds before retrying %r.", delay, tx)
+                            time.sleep(delay)
+                            tx.logger.info("Retrying %r.", tx)
+                    else:
+                        raise
+                finally:
+                    self._set_tx(outer_tx)
+
+        return wrapper
+
+    def _new_tx(self, name: str) -> 'TX':
+        timeout = self.timeout
+        expiration = None if timeout is None else time.time() + timeout
+        return self.TX(name=name, expiration=expiration, logger=self.logger)
+
+    @classmethod
+    def _get_tx(cls) -> Optional['TX']:
+        return getattr(cls.thread_local, 'tx', None)
+
+    @classmethod
+    def _set_tx(cls, tx: 'TX'):
+        cls.thread_local.tx = tx
+
+    @classmethod
+    def get_tx(cls) -> 'TX':
+        tx = cls._get_tx()
+        if tx is None:
+            raise LookupError('No retry transaction active in current thread')
+        else:
+            return tx
+
+    @classmethod
+    def add_context(cls, **kwargs) -> None:
+        cls.get_tx().update(kwargs)
+
+    @classmethod
+    def set_logger(cls, logger: logging.Logger) -> None:
+        cls.get_tx().logger = logger
+
+    class TX(dict):
+
+        def __init__(self, name: str, expiration: float, logger: logging.Logger) -> None:
+            super().__init__()
+            self.name = name
+            self.logger = logger
+            self.expiration = expiration
+
+        def would_expire_after(self, delay: float):
+            return self.expiration is not None and self.expiration < time.time() + delay
+
+        def copy(self, name: str = None):
+            other = type(self)(name=name or self.name, expiration=self.expiration, logger=self.logger)
+            other.update(self)
+            return other
+
+        def log_exception(self, e: BaseException) -> None:
+            if self.logger.isEnabledFor(logging.WARNING):
+                logged_attr_name = f"_{__name__}_already_logged"
+                already_logged = getattr(e, logged_attr_name, False)
+                if already_logged:
+                    self.logger.warning("Exception '%s' occurred in %r. The back trace was logged earlier.", e, self)
+                else:
+                    self.logger.warning("An exception occurred in %r.", self, exc_info=True, stack_info=True)
+                    try:
+                        setattr(e, logged_attr_name, True)
+                    except AttributeError:
+                        self.logger.warning('Failed to mark exception as logged. The back trace may be logged again.')
+
+        def __repr__(self):
+            return f"retry.TX(name='{self.name}', {super().__repr__()})"
+
+    # Overide __new__ to make @retry equivalent to @retry()
```

Comment:
Nit: This should have the `TODO` tag.

Reply:
since this is cookie cutter copied from DSS i'll update it when the DSS updates it.